hudi-common/src/main/java/org/apache/hudi/common/util/queue/BoundedInMemoryExecutor.java

```diff
@@ -168,6 +169,35 @@ public boolean isRemaining() {
   public void shutdownNow() {
     producerExecutorService.shutdownNow();
     consumerExecutorService.shutdownNow();
+    // close queue to force producer stop
+    queue.close();
+  }
+
+  public boolean awaitTermination() {
+    boolean interruptedBefore = Thread.currentThread().isInterrupted();
+    boolean producerTerminated = false;
+    boolean consumerTerminated = false;
+    try {
+      producerTerminated = producerExecutorService.awaitTermination(TERMINATE_WAITING_TIME_SECS, TimeUnit.SECONDS);
+      consumerTerminated = consumerExecutorService.awaitTermination(TERMINATE_WAITING_TIME_SECS, TimeUnit.SECONDS);
+      return producerTerminated && consumerTerminated;
+    } catch (InterruptedException ie) {
+      if (!interruptedBefore) {
+        Thread.currentThread().interrupt();
+        return false;
+      }
+      // if current thread has been interrupted before awaitTermination was called.
```

Comment:
If i understood your intention correctly, you want to give a chance to executors to shutdown properly before proceeding, right? 

You don't need to duplicate code for this you can just call `interrupted` instead of `isInterrupted` and that would clear the interrupted state flag, making sure that following awaitTermination invocations won't fail (line 181, 182)

Reply:
Thanks for your suggestion. Refine the code and make the comment more clear.